main/core/src/mill/eval/Evaluator.scala

```diff
@@ -426,9 +427,33 @@ class Evaluator private (
             catch { case e: Throwable => None }
         } yield (parsed, cached.valueHash)
 
-        val workerCached: Option[Any] = labelledNamedTask.task.asWorker
-          .flatMap { w => synchronized { workerCache.get(w.ctx.segments) } }
-          .collect { case (`inputsHash`, v) => v }
+        val previousWorker = labelledNamedTask.task.asWorker.flatMap { w =>
+          workerCache.synchronized { workerCache.get(w.ctx.segments) }
+        }
+        val workerCached: Option[Any] = previousWorker.collect { case (`inputsHash`, v) => v }
+
+        // Cleanup outdated worker
+        previousWorker
+          .collect {
+            case (hash, v) if v.isInstanceOf[AutoCloseable] && inputsHash != hash => v
+          }
+          .foreach { closableWorkerValue =>
```

Comment:
Instead of a separate  `collect` and `foreach`, have you considered  a single `foreach` with two match inside?
like
```scala
.foreach {
  case (hash, v: AutoCloseable) if inputsHash != hash =>
    // do the things with v
  case _ => // do nothing
}
```

Reply:
Yeah, but it doesn't improve much. I have another idea though...

Reply:
I rewrote it.